pkg/types/finding.go

```diff
@@ -45,3 +51,96 @@ func NewModifiedFinding(f finding, status FindingStatus, statement, source strin
 		Finding:   f,
 	}
 }
+
+// MarshalJSON correctly marshals ModifiedFinding.Finding given the type and `MarshalJSON` functions of struct fields
+func (m *ModifiedFinding) MarshalJSON() ([]byte, error) {
+	var raw struct {
+		Type      FindingType     `json:"Type"`
+		Status    FindingStatus   `json:"Status"`
+		Statement string          `json:"Statement"`
+		Source    string          `json:"Source"`
+		Finding   json.RawMessage `json:"Finding"`
+	}
+	raw.Type = m.Type
+	raw.Status = m.Status
+	raw.Statement = m.Statement
+	raw.Source = m.Source
+
+	// Define a `Finding` type and marshal as a struct of that type.
+	// This is necessary to run the `MarshalJSON` functions on the struct fields.
+	var err error
+	switch val := m.Finding.(type) {
+	case DetectedVulnerability:
+		if raw.Finding, err = json.Marshal(&val); err != nil {
+			return nil, xerrors.Errorf("unable to marshal `DetectedVulnerability` Findings: %w", err)
+		}
+	case DetectedMisconfiguration:
+		if raw.Finding, err = json.Marshal(&val); err != nil {
+			return nil, xerrors.Errorf("unable to marshal `DetectedMisconfiguration` Findings: %w", err)
+		}
+	case DetectedSecret:
+		if raw.Finding, err = json.Marshal(&val); err != nil {
+			return nil, xerrors.Errorf("unable to marshal `DetectedSecret` Findings: %w", err)
+		}
+	case DetectedLicense:
+		if raw.Finding, err = json.Marshal(&val); err != nil {
+			return nil, xerrors.Errorf("unable to marshal `DetectedLicense` Findings: %w", err)
+		}
+	default:
+		return nil, xerrors.Errorf("invalid Finding type: %T", val)
+	}
+
+	return json.Marshal(&raw)
+}
+
+// UnmarshalJSON unmarshals ModifiedFinding given the type and `UnmarshalJSON` functions of struct fields
+func (m *ModifiedFinding) UnmarshalJSON(data []byte) error {
+	raw := struct {
+		Type      FindingType     `json:"Type"`
+		Status    FindingStatus   `json:"Status"`
+		Statement string          `json:"Statement"`
+		Source    string          `json:"Source"`
+		Finding   json.RawMessage `json:"Finding"`
+	}{}
+
+	if err := json.Unmarshal(data, &raw); err != nil {
+		return err
+	}
+
+	m.Type = raw.Type
+	m.Status = raw.Status
+	m.Statement = raw.Statement
+	m.Source = raw.Source
```

Comment:
Can we refactor as below?

```
    type Alias ModifiedFinding
    aux := &struct {
        Finding json.RawMessage `json:"Finding"`
        *Alias
    }{
        Alias: (*Alias)(m),
    }
```

cf. https://github.com/aquasecurity/trivy/blob/5c37361600d922db27dd594b2a80c010a19b3a6e/pkg/fanal/types/package.go#L97-L121

Reply:
Updated in https://github.com/aquasecurity/trivy/pull/7463/commits/06136214854575d0a54c411a8b1add802b24acf1